frontend/src/App.tsx

```diff
@@ -634,12 +636,36 @@ export class App extends PureComponent<Props, State> {
       this.handleOneTimeInitialization(newSessionProto)
     }
 
-    // mainPageName must be a string as we're guaranteed at this point that
+    const config = newSessionProto.config as Config
+    const themeInput = newSessionProto.customTheme as CustomThemeConfig
+    const { currentPageScriptHash } = this.state
+    const newPageScriptHash = newSessionProto.pageScriptHash
+
+    // mainPage must be a string as we're guaranteed at this point that
     // newSessionProto.appPages is nonempty and has a truthy pageName.
     // Otherwise, we'd either have no main script or a nameless main script,
     // neither of which can happen.
-    const mainPageName = newSessionProto.appPages[0]?.pageName as string
-    const { currentPageName } = this.state
+    const mainPage = newSessionProto.appPages[0] as AppPage
+    // We're similarly guaranteed that newPageName will be found / truthy
+    // here.
+    const newPageName = newSessionProto.appPages.find(
+      p => p.pageScriptHash === newPageScriptHash
+    )?.pageName as string
+    const viewingMainPage = newPageScriptHash === mainPage.pageScriptHash
+
+    const baseUriParts = this.getBaseUriParts()
+    if (baseUriParts) {
+      const { basePath } = baseUriParts
+      const queryString = this.getQueryString()
+
+      const qs = queryString ? `?${queryString}` : ""
```

Comment:
can line 661 and 659 be combined to be 
`const queryString = this.getQueryString() ? `?${queryString}` : ""`

Reply:
You can't use `queryString` on the RHS before it's actually defined, and I think calling `this.getQueryString()` twice would be worse